### PR TITLE
Try to fix some docking memory issues

### PR DESCRIPTION
**Pull request by apayne97**
## Description
There seem to be some memory accumulation errors as mentioned in #212 during docking runs. It's not clear when / why this happens. After spending some time trying to debug the issue, instead this PR just adds some code for debugging, and for running docking as a job array to minimize the number of tasks each individual job needs to perform.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] add function to split up multi-ligand sdf into chunks of sdf files `asapdiscovery.data.split_sdf.py`
  - [x] add cluster script to be run as a job array
  - [x] fix some minor bugs in the run_docking_oe script and function
  - [x] add datetime checks to see how long the various steps take
  - [x] add log_name param to `run_docking_oe.py` to use as basename string for high-level log files
  - [x] change import of ML code to only happen if a GAT model is passed
  - [x] add pebble to dependencies
  - [x] get tests working

## Questions
- [x] should I add back the pickle reading and dataframe generating? I'm leaning against it
- I decided to add it back! it didn't seem to require much time and it is kinda convenient to have it in this script

## Status
- [x] Ready to go

### CODE REVIEW

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (kaminow) [inline asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py:54]: is there a reason these imports aren't at the top of the script?

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (kaminow) [inline asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py:51]:
```suggestion
    if remainder:
```

**Review by kaminow — APPROVED** [review]
couple minor suggestions, but I think this looks great! would probably be good to get a quick check from another set of eyes though

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (apayne97) [reply asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py:51]: LOL

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (apayne97) [reply asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py]: nope. fixed here: https://github.com/choderalab/asapdiscovery/pull/213/commits/cbe5b792b5efd5a9315ae47b8f0f01748815df31

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py:51]: I could be missing something but we arn't doing anything with `remainder`?

File: devtools/conda-envs/asapdiscovery.yaml
Comment (hmacdope) [inline devtools/conda-envs/asapdiscovery.yaml:58]: These are from #223 ?

**Review by hmacdope — CHANGES_REQUESTED** [review]

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (JenkeScheen) [inline asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:358]: is this intended? @hmacdope is this a good idea?

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (JenkeScheen) [reply asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py:51]: right, `remainder` still has to be written out to file as well. It's the tail end of mols if `len(mols)` is not divisible by `chunk_size`

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (JenkeScheen) [inline asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py:52]: if possible might be good to state the file name of the `remainder` file, this will always be smaller than the rest of the files, so handy for debugging.

File: asapdiscovery-docking/asapdiscovery/docking/cluster_scripts/run_fragalysis_retrospective_array.sh
Comment (JenkeScheen) [inline asapdiscovery-docking/asapdiscovery/docking/cluster_scripts/run_fragalysis_retrospective_array.sh:13]: this will break for other users of course, perhaps best to put in a comment to point the users to their own conda envs

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (apayne97) [reply asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py:51]: since i'm using the ceiling function, I actually am also saving out the remainder. However I've decided to try to rewrite it in a way that makes it more obvious

Comment (apayne97) [reply asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py]: done in https://github.com/choderalab/asapdiscovery/pull/213/commits/c7ab339ca6eabf1f73a7ca3c2372680aa417f483

File: asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py
Comment (apayne97) [reply asapdiscovery-data/asapdiscovery/data/scripts/split_sdf.py]: done here: https://github.com/choderalab/asapdiscovery/pull/213/commits/c7ab339ca6eabf1f73a7ca3c2372680aa417f483

File: asapdiscovery-docking/asapdiscovery/docking/cluster_scripts/run_fragalysis_retrospective_array.sh
Comment (apayne97) [reply asapdiscovery-docking/asapdiscovery/docking/cluster_scripts/run_fragalysis_retrospective_array.sh:13]: this is true for all the cluster scripts...i'll make a separate issue for this, maybe there's a better way to keep these in the repo? i'd appreciate keeping them

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (apayne97) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:358]: this was from #218 and #224; I don't think the default should be to have a timer, since the expected time to completion might be different for every kind of docking run? (and, i.e. will be 10x different if you use the Omega and clash relax flags vs default) what do you think?

File: asapdiscovery-docking/asapdiscovery/docking/cluster_scripts/run_fragalysis_retrospective_array.sh
Comment (apayne97) [reply asapdiscovery-docking/asapdiscovery/docking/cluster_scripts/run_fragalysis_retrospective_array.sh:13]: Added issue here https://github.com/choderalab/asapdiscovery/issues/230

File: devtools/conda-envs/asapdiscovery.yaml
Comment (apayne97) [reply devtools/conda-envs/asapdiscovery.yaml:58]: you added these in https://github.com/choderalab/asapdiscovery/commit/adc1150635b9740bf2a18e753c6157caacd2fe97

Comment (apayne97) [reply devtools/conda-envs/asapdiscovery.yaml:58]: i added the pebble changes because of #218

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:358]: I also had some issues with it being too short on my setup and will differ wildly depending on hardware. I agree its hard to judge exactly how long it should take. TBH I think changing to a large value (say 300 sec) is probably better than disabling it altogether.

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (hmacdope) [inline asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:434]: Why do we need this `global` ? I might be missing something but do we access it from global scope ever? Also seems to be defined in two places.  If possible better to define once (non-global) in `main` pass as an argument to whatever you need

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (hmacdope) [inline asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:480]: Good idea to move, thanks. :)

File: devtools/conda-envs/asapdiscovery.yaml
Comment (hmacdope) [reply devtools/conda-envs/asapdiscovery.yaml:58]: Ah yep ok.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Quick queries but thanks for the improvements. :)

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (JenkeScheen) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:358]: I agree that we should set an overly long timeout just to prevent jobs from hanging until eternity.

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (apayne97) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:434]: @hmacdope as discussed i'm enabling this now

Comment (apayne97) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py]: done here: https://github.com/choderalab/asapdiscovery/pull/213/commits/021ee5a4692764931e4ccbfe476390e98dd15508

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (apayne97) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py]: changed to 300 seconds here: https://github.com/choderalab/asapdiscovery/pull/213/commits/aab600f4bf0e756aa1865fa0e58952fb9f95efc5

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py
Comment (apayne97) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:538]: this is using the pathlib `/` operator! output_dir is a Path object.

Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:538]: Wow had no idea that was possible.  Great!

**Review by hmacdope — APPROVED** [review]
LGTM, merge when ready @apayne97

**Comment by apayne97** [comment]
@hmacdope shocked that you responded so quickly, but I guess it's early enough for me that it isn't *too* late for you. go to bed though :P <3

**Pull request merged**
